src/sqlfluff/rules/L003.py

```diff
@@ -416,113 +466,76 @@ def _process_current_line(
         - When passed the *final* segment in the entire parse tree (which may
           not be a newline)
         """
+        res = memory.line_summaries
         this_line_no = max(res.keys())
-        this_line = res.pop(this_line_no)
+        this_line: _LineSummary = res.pop(this_line_no)
         self.logger.debug(
             "Evaluating line #%s. %s",
             this_line_no,
             # Don't log the line or indent buffer, it's too noisy.
-            self._strip_buffers(this_line),
+            this_line,
         )
-        trigger_segment = memory["trigger"]
 
-        # Is this line just comments? (Disregard trailing newline if present.)
-        check_comment_line = this_line["line_buffer"]
-        if check_comment_line and all(
-            seg.is_type(
-                "whitespace", "comment", "indent"  # dedent is a subtype of indent
-            )
-            for seg in check_comment_line
-        ):
+        if this_line.is_comment_line:
             # Comment line, deal with it later.
-            memory["comment_lines"].append(this_line_no)
+            memory.comment_lines.append(this_line_no)
             self.logger.debug("    Comment Line. #%s", this_line_no)
             return LintResult(memory=memory)
 
-        # Is it a hanging indent?
-        # Find last meaningful line indent.
-        last_code_line = None
-        for k in sorted(res.keys(), reverse=True):
-            if any(seg.is_code for seg in res[k]["line_buffer"]):
-                last_code_line = k
-                break
-
-        if len(res) > 0 and last_code_line:
-            last_line_hanger_indent = res[last_code_line]["hanging_indent"]
+        last_line = _find_last_meaningful_line(res)
+        if len(res) > 0 and last_line:
             # Let's just deal with hanging indents here.
             if (
                 # NB: Hangers are only allowed if there was content after the last
                 # indent on the previous line. Otherwise it's just an indent.
-                this_line["indent_size"] == last_line_hanger_indent
+                this_line.indent_size == last_line.hanging_indent
                 # Or they're if the indent balance is the same and the indent is the
                 # same AND the previous line was a hanger
                 or (
-                    this_line["indent_size"] == res[last_code_line]["indent_size"]
-                    and this_line["indent_balance"]
-                    == res[last_code_line]["indent_balance"]
-                    and last_code_line in memory["hanging_lines"]
+                    this_line.indent_size == last_line.indent_size
+                    and this_line.anchor_indent_balance
+                    == last_line.anchor_indent_balance
```

Comment:
`anchor_indent_balance` replaces `indent_balance` in many contexts - the variable was holding two different concepts which were hard to reason about.

`anchor_indent_balance`  => Is the indent_balance state at the anchor element (first el after a lines starting whitespace) - we do most calculations with the current also using this.
`indent_balance` => The indent balance after the whole line has been calculated/considered

Reply:
This is a good change - very clearly explained 👍